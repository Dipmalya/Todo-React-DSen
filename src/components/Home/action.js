import { ADD_TODO } from '../../config/actionTypes';

/**
 * Method to add todo
 * @param {todo} value 
 */
export const addTodoItem = value => {
    return (dispatch) => {
        dispatch(addTodoDispatcher(value))
    };
}

const addTodoDispatcher = (value) => ({
    type: ADD_TODO,
    payload: value
});