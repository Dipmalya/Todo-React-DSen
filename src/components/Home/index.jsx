import React, { Component } from "react";
import TodoPanel from "../TodoPanel/container";

const { Home: data } = CONFIG;

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentView: "ALL",
    };
  }

  renderHeader = () => (
    <div className="bg-dark py-4">
      <div className="text-white text-center container">
        <h1>{data.heading}</h1>
      </div>
    </div>
  );

  handleBlurInput = (event) => {
    const { target: { name, value } } = event;
    this.setState({ [name]: value });
  }

  handleAddTodo = () => {
    const { todoText } = this.state;
    this.props.addTodoItem(todoText);
    this.setState({ todoText: '' });
  }

  renderInput = () => {
    const { input: { todoTextInput, todoBtnInput } } = data;
    const { text: btnText, className, iconClassName } = todoBtnInput;
    const { todoText } = this.state;
    return (
      <>
        <input
          value={todoText}
          {...todoTextInput}
          onBlur={this.handleBlurInput}
        />
        <button 
          className={className}
          disabled={!todoText}
          onClick={this.handleAddTodo}
        >
          {btnText}
          <i className={iconClassName}/>
        </button>
      </>
    );
  };

  handleViewToggle = (value) => {
    this.setState({ currentView: value });
  };

  render() {
    const { currentView } = this.state;
    const { todoList } = this.props;
    return (
      <div>
        {this.renderHeader()}
        <div className="d-flex col-md-4 offset-md-4 mt-md-5 mt-4 form-group">
          {this.renderInput()}
        </div>
        <TodoPanel
          todoList={todoList}
          currentView={currentView}
          toggleView={this.handleViewToggle}
        />
      </div>
    );
  }
}

export default Home;
