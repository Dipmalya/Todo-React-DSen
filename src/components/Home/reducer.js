import { ADD_TODO, DELETE_TODO, TOGGLE_TODO } from '../../config/actionTypes';
import { v4 as uuidv4 } from "uuid";

const { COMPLETED, ACTIVE } = CONSTANT;

export const todoList = (state = [], {type, payload}) => {
    switch(type) {
        case ADD_TODO: {
            return [ ...state, { id: uuidv4(), task: payload, status: ACTIVE } ]
        };
        case DELETE_TODO: {
            return state.filter(item => item.id !== payload);
        }
        case TOGGLE_TODO: {
            return state.map(todo => {
                const { status: currentStatus } = todo;
                return (
                    todo.id === payload ? 
                    { ...todo, status: currentStatus === ACTIVE ? COMPLETED : ACTIVE } : todo
                )
            });
        }
        default: {
            return state;
        }
    }
}