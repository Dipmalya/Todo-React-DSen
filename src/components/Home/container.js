import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'
import Home from './index';

import { addTodoItem } from './action';

const mapStateToProps = ({ todoList }) => ({
    todoList
});

const mapDispatchToProps = (dispatch) => ({
    addTodoItem: bindActionCreators(addTodoItem, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);