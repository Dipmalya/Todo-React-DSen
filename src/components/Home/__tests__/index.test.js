import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Home from '../index';

jest.mock("../../TodoPanel/container");

const props = {
    todoList: [
        {
            id: 'bda37a11-6a12-40ec-9365-331a516194ea',
            task: 'Do next',
            status: 'ACTIVE'
        }
    ],
    addTodoItem: jest.fn()
};
const mockEvent = {
    target: {
        name: 'todoText',
        value: 'mockValue'
    }
};

describe('Home Component', () => {
    const wrapper = mount(<Home {...props} />);

    describe('Test suite to check component rendering', () => {
        it('Test to check Home component gets rendered successfully', () => {
            expect(wrapper.exists()).toBe(true);
        });
    
        it('Test to check snapshot of the component is validated', () => {
            const component = renderer.create(<Home {...props} />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    
        it('Test to check input field gets rendered', () => {
            const dummyTodoInput = wrapper.find('input[id="task"]');
            expect(dummyTodoInput.exists()).toBe(true);
        });
    
        it('Test to check Add button gets rendered', () => {
            const dummyTodoBtn = wrapper.find('button[className="btn btn-info ml-3"]');
            expect(dummyTodoBtn.exists()).toBe(true);
        });
    });

    describe('Test suite to check UI functionality', () => {
        it('Test to check todo is added on inputting value', () => {
            const dummyTodoBtn = wrapper.find('input[id="task"]');
            dummyTodoBtn.props().onChange(mockEvent);
            expect(wrapper.instance().state.todoText).toEqual('mockValue');
        });
    
        it('Test to check action method is called on button click', () => {
            const mockAddTodo = jest.fn();
            wrapper.setProps({ addTodoItem: mockAddTodo });
            const dummyTodoBtn = wrapper.find('button[className="btn btn-info ml-3"]');
            dummyTodoBtn.props().onClick();
            expect(mockAddTodo).toBeCalled();
        });

        it('Test to check user views ALL todo by default', () => {
            expect(wrapper.instance().state.currentView).toEqual('ALL');
        });
    
        it('Test to check validation if todo exists and error is thrown', () => {
            const sameTodo = {
                target: {
                    name: 'todoText',
                    value: 'Do next'
                }
            };
            const dummyTodoBtn = wrapper.find('input[id="task"]');
            dummyTodoBtn.props().onBlur(sameTodo);
            expect(wrapper.instance().state.error).toBeTruthy();
        });

        it('Test to check error is disabled on focus of input', () => {
            const dummyTodoBtn = wrapper.find('input[id="task"]');
            dummyTodoBtn.props().onFocus();
            expect(wrapper.instance().state.error).toEqual('');
        });
    })
});
