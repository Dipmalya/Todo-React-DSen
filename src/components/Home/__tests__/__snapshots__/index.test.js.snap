// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Test suite to check component rendering Test to check snapshot of the component is validated 1`] = `
<div>
  <div
    className="bg-dark py-4"
  >
    <div
      className="text-white text-center container"
    >
      <h1>
        ToDo List
      </h1>
    </div>
  </div>
  <div
    className="d-flex col-md-4 offset-md-4 mt-md-5 mt-4"
  >
    <input
      className="form-control"
      id="task"
      name="todoText"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Add todo.."
      type="text"
      value=""
    />
    <button
      className="btn btn-info ml-3"
      disabled={true}
      onClick={[Function]}
    >
      ADD
      <i
        className="ml-2 fa fa-arrow-right"
      />
    </button>
  </div>
  <div
    className="col-md-4 offset-md-4 error-text"
  />
</div>
`;
